File: homework/thinhbka/lesson-7-twopointer-heap/1004. Max Consecutive Ones III.py

# Given a binary array nums and an integer k, return the maximum number of consecutive 1's in the array if you can flip at most k 0's.
# Input: nums = [1,1,1,0,0,0,1,1,1,1,0], k = 2
# Output: 6
# Explanation: [1,1,1,0,0,1,1,1,1,1,1]
# Bolded numbers were flipped from 0 to 1. The longest subarray is underlined.
# [1,1,1,0,0,0,1,1,1,1,0]
        #    i
class Solution:
    def longestOnes(self, nums, k) -> int:
        ans  = nums[0]
        left = 0
        cnt = k
        for i in range(len(nums)):
            if nums[i] == 1:
                ans = max(ans,i-left+1)
                print(ans)
            else:
                cnt-=1
                while nums[left]!=0:
                    left+=1
                left+=1
                cnt+=1
        return ans
Solution().longestOnes([1,1,1,0,0,0,1,1,1,1,0],2)
print("")
Solution().longestOnes([0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1],3)

File: homework/tai_nguyen/lesson10/1004_max_consecutive_ones_iii.ts
function longestOnes(nums: number[], k: number): number {
  let l = 0,
    r = 0;
  let zeroCount = nums[0] === 0 ? 1 : 0;
  let res = 0;

  while (r < nums.length) {
    if (zeroCount <= k) {
      res = Math.max(res, r - l + 1);
      r++;
      if (r < nums.length) {
        if (nums[r] === 0) {
          zeroCount++;
        }
      }
    } else {
      if (nums[l] === 0) {
        zeroCount--;
      }
      l++;
    }
  }

  return res;
}
